File: Haskell/2020 Septiembre/01_Hecho/AVL.hs
{------------------------------------------------------------------------------
 - Student's name:
 -
 - Student's group:
 -----------------------------------------------------------------------------}

module AVL 
  ( 
    Weight
  , Capacity
  , AVL (..)
  , Bin
  , emptyBin
  , remainingCapacity
  , addObject
  , maxRemainingCapacity
  , height
  , nodeWithHeight
  , node
  , rotateLeft
  , addNewBin
  , addFirst
  , addAll
  , toList
  , linearBinPacking
  , seqToList
  , addAllFold
  ) where

type Capacity = Int
type Weight= Int

data Bin = B Capacity [Weight] 

data AVL = Empty | Node Bin Int Capacity AVL AVL deriving Show


emptyBin :: Capacity -> Bin
emptyBin c = (B c [])

remainingCapacity :: Bin -> Capacity
remainingCapacity (B c xs) = c

addObject :: Weight -> Bin -> Bin
addObject w (B c xs)
  | w <= c = (B (c-w) (w:xs))
  | otherwise = error "No cabe el objeto en el cubo"

maxRemainingCapacity :: AVL -> Capacity
maxRemainingCapacity Empty = 0
maxRemainingCapacity (Node (B c xs) altura capacidad izq der) = capacidad

height :: AVL -> Int
height Empty = 0
height (Node (B c xs) altura capacidad izq der) = altura


 
nodeWithHeight :: Bin -> Int -> AVL -> AVL -> AVL
nodeWithHeight (B c xs) altura Empty Empty = ( Node (B c xs) altura c Empty Empty )

nodeWithHeight (B c xs) altura Empty (Node (B cd xsd) alturad capacidadd izqd derd)
  | c >= capacidadd = ( Node (B c xs) altura c Empty (Node (B cd xsd) alturad capacidadd izqd derd) )
  | otherwise = ( Node (B c xs) altura capacidadd Empty (Node (B cd xsd) alturad capacidadd izqd derd) )

nodeWithHeight (B c xs) altura (Node (B ci xsi) alturai capacidadi izqi deri) Empty
  | c >= capacidadi = ( Node (B c xs) altura c (Node (B ci xsi) alturai capacidadi izqi deri) Empty )
  | otherwise = ( Node (B c xs) altura capacidadi (Node (B ci xsi) alturai capacidadi izqi deri) Empty )

nodeWithHeight (B c xs) altura (Node (B ci xsi) alturai capacidadi izqi deri) (Node (B cd xsd) alturad capacidadd izqd derd)
  | (c >= capacidadi) && (c >= capacidadd) = ( Node (B c xs) altura c (Node (B ci xsi) alturai capacidadi izqi deri) (Node (B cd xsd) alturad capacidadd izqd derd) )
  | (capacidadi >= c) && (capacidadi >= capacidadd) = ( Node (B c xs) altura capacidadi (Node (B ci xsi) alturai capacidadi izqi deri) (Node (B cd xsd) alturad capacidadd izqd derd) )
  | otherwise = (Node (B c xs) altura capacidadi (Node (B ci xsi) alturai capacidadd izqi deri) (Node (B cd xsd) alturad capacidadd izqd derd) )


node :: Bin -> AVL -> AVL -> AVL
node (B c xs) Empty Empty = nodeWithHeight (B c xs) 1 Empty Empty
node (B c xs) Empty (Node (B cd xsd) alturad capacidadd izqd derd) = nodeWithHeight (B c xs) (alturad+1)  Empty (Node (B cd xsd) alturad capacidadd izqd derd)
node (B c xs) (Node (B ci xsi) alturai capacidadi izqi deri) Empty = nodeWithHeight (B c xs) (alturai+1) (Node (B ci xsi) alturai capacidadi izqi deri) Empty
node (B c xs) (Node (B ci xsi) alturai capacidadi izqi deri) (Node (B cd xsd) alturad capacidadd izqd derd)
  | alturai >= alturad = nodeWithHeight (B c xs) (alturai+1) (Node (B ci xsi) alturai capacidadi izqi deri) (Node (B cd xsd) alturad capacidadd izqd derd)
  | otherwise = nodeWithHeight (B c xs) (alturad+1) (Node (B ci xsi) alturai capacidadi izqi deri) (Node (B cd xsd) alturad capacidadd izqd derd)


rotateLeft :: Bin -> AVL -> AVL -> AVL
rotateLeft (B c xs) izq (Node (B cd xsd) alturad capacidadd izqd derd) = root
  where
    nizq = node (B c xs) izq izqd
    root = node (B cd xsd) nizq derd 


addNewBin :: Bin -> AVL -> AVL
addNewBin (B c xs) Empty = (Node (B c xs) 1 c Empty Empty)
addNewBin (B c xs) (Node (B cn xsn) altura capacidad izq der) 
  | (height der) - (height izq) > 1 = rotateLeft (B c xs) izq (addNewBin (B c xs) der)
  | otherwise = (Node (B cn xsn) altura capacidad izq (addNewBin (B c xs) der))

 
addFirst :: Capacity -> Weight -> AVL -> AVL
addFirst capacidadrest pesoagregar Empty = addNewBin (B (capacidadrest-pesoagregar) [pesoagregar]) Empty
addFirst capacidadrest pesoagregar (Node (B c xs) altura capacidad izq der)
  | capacidadrest < pesoagregar = addNewBin (B (capacidadrest-pesoagregar) [pesoagregar]) (Node (B c xs) altura capacidad izq der)
  | maxRemainingCapacity izq >= pesoagregar =  (Node (B c xs) altura capacidad (addFirst capacidadrest pesoagregar izq) der)
  | maxRemainingCapacity (Node (B c xs) altura capacidad izq der) >= pesoagregar = (Node ( addObject (pesoagregar) (B c xs) ) altura (capacidad-pesoagregar) izq der)
  | otherwise = (Node (B c xs) altura capacidad izq (addFirst capacidadrest pesoagregar der) )

addAll:: Capacity -> [Weight] -> AVL
addAll cap (x:xs) = addAllRec cap (x:xs) Empty

addAllRec :: Capacity -> [Weight] -> AVL -> AVL
addAllRec cap [] avl = avl
addAllRec cap (x:xs) avl = addAllRec cap xs (addFirst cap x avl)  

toList :: AVL -> [Bin]
toList Empty = []
toList (Node bin a c izq der) = toList izq ++ [bin] ++ toList der

{-
	SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
 -}

data Sequence = SEmpty | SNode Bin Sequence deriving Show  

linearBinPacking:: Capacity -> [Weight] -> Sequence
linearBinPacking _ _ = undefined

seqToList:: Sequence -> [Bin]
seqToList _ = undefined

addAllFold:: [Weight] -> Capacity -> AVL 
addAllFold _ _ = undefined



{- No modificar. Do not edit -}

objects :: Bin -> [Weight]
objects (B _ os) = reverse os

  
instance Show Bin where
  show b@(B c os) = "Bin("++show c++","++show (objects b)++")"

File: Java/2020 Septiembre/01_Hecho/src/AVL.java
/**
 * Student's name:
 *
 * Student's group:
 */

import dataStructures.list.ArrayList;
import dataStructures.list.List;
import dataStructures.list.LinkedList;

import java.util.Iterator;


class Bin {
    private int remainingCapacity; // capacity left for this bin
    private List<Integer> weights; // weights of objects included in this bin

    public Bin(int initialCapacity) {
        // todo

        remainingCapacity = initialCapacity;
        weights = new ArrayList<>();

    }

    // returns capacity left for this bin
    public int remainingCapacity() {
        // todo
        return remainingCapacity;
    }

    // adds a new object to this bin
    public void addObject(int weight) {
        // todo

        if (weight > remainingCapacity)
        {
            throw new RuntimeException("El cubo no cabe");
        }

        weights.append(weight);
        remainingCapacity = remainingCapacity - weight;

    }

    // returns an iterable through weights of objects included in this bin
    public Iterable<Integer> objects() {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        sb.append(remainingCapacity);
        sb.append(", ");
        sb.append(weights.toString());
        sb.append(")");
        return sb.toString();
    }
}

// Class for representing an AVL tree of bins
public class AVL {
    static private class Node {
        Bin bin; // Bin stored in this node
        int height; // height of this node in AVL tree
        int maxRemainingCapacity; // max capacity left among all bins in tree rooted at this node
        Node left, right; // left and right children of this node in AVL tree

        // recomputes height of this node
        void setHeight() {
            // todo
            if(left == null && right == null)
            {
                height = 1;
            } else if (left == null && right != null)
            {

                height = right.height+1;

            } else if (left != null & right == null)
            {
                height = left.height+1;
            } else
            {
                height = Math.max(left.height, right.height)+1;
            }

        }

        // recomputes max capacity among bins in tree rooted at this node
        void setMaxRemainingCapacity() {
            // todo

            if(left == null && right == null)
            {
                maxRemainingCapacity = bin.remainingCapacity();
            } else if (left != null && right == null)
            {
                maxRemainingCapacity = Math.max(bin.remainingCapacity(), left.maxRemainingCapacity);
            } else if ( left == null && right != null)
            {
                maxRemainingCapacity = Math.max(bin.remainingCapacity(), right.maxRemainingCapacity);
            } else
            {
                maxRemainingCapacity = Math.max( bin.remainingCapacity(), Math.max(right.maxRemainingCapacity, left.maxRemainingCapacity) );
            }

        }

        // left-rotates this node. Returns root of resulting rotated tree
        Node rotateLeft() {
            // todo

            Node rootBueno = right; //root a devolver
            this.right = rootBueno.left;
            rootBueno.left = this;

            this.setMaxRemainingCapacity();
            this.setHeight();

            rootBueno.setHeight();
            rootBueno.setMaxRemainingCapacity();

            return rootBueno;
        }
    }

    private static int height(Node node) {
        // todo
        return node.height;
    }

    private static int maxRemainingCapacity(Node node) {
        // todo
        return node.maxRemainingCapacity;
    }

    private Node root; // root of AVL tree

    public AVL() {
        this.root = null;
    }

    // adds a new bin at the end of right spine.
    private void addNewBin(Bin bin) {
        // todo

        root = addnewBinRec(bin, root);

    }

    private Node addnewBinRec (Bin bin, Node nodo)
    {
        if(nodo == null)
        {
            Node nuevoNodo = new Node();
            nuevoNodo = new Node();
            nuevoNodo.bin = bin;
            nuevoNodo.maxRemainingCapacity = bin.remainingCapacity();
            nuevoNodo.height = 1;

            return nuevoNodo;

        } else
        {
            nodo.right = addnewBinRec(bin, nodo.right);

            if(nodo.left != null && nodo.right != null)
            {
                if(nodo.left.height+1 < nodo.right.height)
                {
                    nodo.rotateLeft();
                }
            }
            nodo.setMaxRemainingCapacity();
            nodo.setHeight();

            return nodo;
        }


    }

    // adds an object to first suitable bin. Adds
    // a new bin if object cannot be inserted in any existing bin
    public void addFirst(int initialCapacity, int weight) {
        // todo
        addFirstRec(initialCapacity, weight, root);
    }

    public void addFirstRec(int initialCapacity, int weight, Node nodo) {
        // todo

        if(nodo == null || weight > nodo.maxRemainingCapacity)
        {
            Bin bin = new Bin (initialCapacity);
            bin.addObject(weight);
            addNewBin(bin);
        } else if (nodo.left != null && nodo.left.maxRemainingCapacity >= weight)
        {
            addFirstRec(initialCapacity, weight, nodo.left);
        } else if (nodo.bin.remainingCapacity() >= weight)
        {
            nodo.bin.addObject(weight);
        } else
        {
            addFirstRec(initialCapacity, weight, nodo.right);
        }


    }

    public void addAll(int initialCapacity, int[] weights) {
        // todo

        for(int i = 0; i<weights.length; i++)
        {
            addFirst(initialCapacity, weights[i]);
        }

    }

    public List<Bin> toList() {
        // todo

        List<Bin> l = new ArrayList<>();
        enOrden(l, root);

        return l;
    }

    private void enOrden (List<Bin> l, Node nodo)
    {

        if(nodo != null)
        {
            System.out.println(nodo.bin);
            if(nodo.left != null)
            {
                enOrden(l, nodo.left);
            }

            l.append(nodo.bin);

            if(nodo.right != null)
            {
                enOrden(l, nodo.right);
            }
        }

    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        stringBuild(sb, root);
        sb.append(")");
        return sb.toString();
    }

    private static void stringBuild(StringBuilder sb, Node node) {
        if(node==null)
            sb.append("null");
        else {
            sb.append(node.getClass().getSimpleName());
            sb.append("(");
            sb.append(node.bin);
            sb.append(", ");
            sb.append(node.height);
            sb.append(", ");
            sb.append(node.maxRemainingCapacity);
            sb.append(", ");
            stringBuild(sb, node.left);
            sb.append(", ");
            stringBuild(sb, node.right);
            sb.append(")");
        }
    }
}

class LinearBinPacking {
    public static List<Bin> linearBinPacking(int initialCapacity, List<Integer> weights) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }
	
	public static Iterable<Integer> allWeights(Iterable<Bin> bins) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;		
	}
}